### PR TITLE
Add webgl draw buffers support

### LINKED ISSUE #142: support for WEBGL_draw_buffers?
How hard would it be to get the https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_draw_buffers extension working?

### PR DESCRIPTION
**Pull request by robertleeplummerjr**
Closes #142 

The unit test I ported from webgl2, to webgl1, to headlessgl, and then as well to google angle, to see how is would hold up across layers.  They are here: https://gist.github.com/robertleeplummerjr/3313587571dd2a8fa3c3907e451b5564

### CODE REVIEW

**Comment by robertleeplummerjr** [comment]
Will resolve this AM.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "gl",
-  "version": "4.1.1",
+  "version": "4.2.0",
```
[the remark this answers is not in the document]
Comment (robertleeplummerjr) [reply package.json]: Rolled back. Ty for clarifying.

File: webgl.js [hunk 2]
```diff
@@ -286,8 +358,65 @@ function precheckFramebufferStatus (framebuffer) {
     return gl.FRAMEBUFFER_UNSUPPORTED
   }
 
-  if (!colorAttachment) {
-    return gl.FRAMEBUFFER_INCOMPLETE_MISSING_ATTACHMENT
+  var colorAttachments
+  var colorAttachment0 = attachments[gl.COLOR_ATTACHMENT0]
+  if (webgl_draw_buffers) {  // eslint-disable-line
+    var colorAttachment1 = attachments[webgl_draw_buffers.COLOR_ATTACHMENT1_WEBGL] // eslint-disable-line
```
[the remark this answers is not in the document]
Comment (robertleeplummerjr) [reply webgl.js]: When I would run `npm test` this lint would prevent unit tests from running.

File: test/multiple-render-targets.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
+'use strict'
+
+var createContext = require('../index')
+var tape = require('tape')
+tape('multiple-render-targets', function (t) {
```
[the remark this answers is not in the document]
Comment (robertleeplummerjr) [reply test/multiple-render-targets.js]: This was erroneous on my part and being new to your test runner.  Added `t.end`.  Ty for finding!

File: test/multiple-render-targets.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      gl.activeTexture(gl.TEXTURE0 + i)
+      gl.bindTexture(gl.TEXTURE_2D, texture)
+    }
+    gl.uniform1iv(gl.getUniformLocation(program, 'tex[0]'), [0, 1, 2, 3])
+    gl.drawArrays(gl.POINTS, 0, 1)
+  }
+
+  function toPixels () {
+    var bytes = new Uint8Array(width * height * 4)
+    gl.var(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, bytes)
```
[the remark this answers is not in the document]
Comment (robertleeplummerjr) [reply test/multiple-render-targets.js]: Again another failure on my part.  I think I was in such excitement about seeing the multi render target image test that I had for both webgl1 and webgl2 outputting, that I must have typed this by accident.  I apologize or the inconvenience and my overly-zealous behavior resulting in utter confusion.

The updated revision has a comment that, when un-commented, should create a file named "multiple-render-targets.ppm" in the root directory of this project that resembles this:

multiple-render-targets

**Comment by robertleeplummerjr** [comment]
I started to look at failures from enabling "extensions_webgl-draw-buffers" from conformance tests, and we have a bit to go to satisfy those, fyi.

As far as the other failures, I only get the following failures:
* glsl_variables_glsl-built-ins
 * not ok 5762 should be green
 * not ok 5763 should be green
* not ok 8026 no plan foundegative level

But the tests to seem to die early with:
```
⨯ fail  3
error Command failed with exit code 1.
```

**Comment by dhritzkiv** [comment]
I'm getting errors with the `multiple-render-targets` test:

```
not ok 8306 Only 45000 are not 0, expected 144000
```

in multiple-render-targets.js, `outWidth` (`300`) * `ouputHeight` (`150`) = `45000`, so that makes some sense. How did you arrive at the expectation of `144000`? Should it may be be `180000` (45000 pixels * 4 channels), even though ?

**Comment by dhritzkiv** [comment]
Edit: the test failures may be the result of the wrong results: after outputting the file, I notice that it's not working on my machine:
screen shot 2019-02-26 at 12 59 42

I wonder how to debug this issue…

**Comment by robertleeplummerjr** [comment]
Just reproduced, will have a fix momentarily.

**Comment by robertleeplummerjr** [comment]
It was the inclusion of the `gl.NONE`, so pulled that back.  Test is now passing locally.

**Comment by robertleeplummerjr** [comment]
The size of the graphical output was simply me trying to find one of the simplest outputs that I hadn't written so to re-emulate.  The original "test" was from here: https://stackoverflow.com/questions/51793336/webgl-2-0-multiple-output-textures-from-the-same-program/51798078#51798078 , which was just me counting the pixels, and yes they are `width * height * 4`, but there is some offset from the edges.  I would be very welcome to simplifying the test, but I think I just welcomed getting the PR opened asap.

**Comment by dhritzkiv** [comment]
Fortunately I am now getting this output: 
image.
Interestingly, there are no black bars.

Further, I am now getting this error for the test:
```
not ok 1 Only 168750 are not 0, expected 144000
```

**Comment by robertleeplummerjr** [comment]
Interesting! I wonder if foreground or background is being lost?  How far do we need to go to get this merged in?  Does it have to be full on unit tests compliant, or can we go in stages?

**Comment by dhritzkiv** [comment]
Doing some basic math:
300*150*4 (w * h * channels) = 180000 (possible expected value?)
180000 - 168750 (actual result) = 11250 (difference)
11250 - (350*150/4) = 0

…I don't how that's significant (or if it's just coincidence), but I would guess that one of the colour's channels values is 0?

Are you getting this result when you run the tests? (btw - you can write `tape.only(` instead of `tape(` to run only the multiple-render-targets test)

**Comment by robertleeplummerjr** [comment]
If I may propose: This unit test is too complex.  We just need a few pixels to know the output, in-fact, was from multi targets.  We can then start to address bugs as they are found, like this one.

**Comment by robertleeplummerjr** [comment]
That is as well a good point, perhaps we should ensure none of them are zero, before checking if they are zero 😛

**Comment by dhritzkiv** [comment]
> How far do we need to go to get this merged in? Does it have to be full on unit tests compliant, or can we go in stages?

Unfortunately for us (but fortunate for the project) we can have no regression in the tests in order to land something in master and push a release. Which means all the tests must pass as the current pass rate is 100%.

Fortunately, it's only a few suites that are failing, and the source of the errors likely lies in a single point of the new code. Find that source, and the tests will likely pass.

**Comment by robertleeplummerjr** [comment]
That isn't what I mean, of course this PR cannot introduce new bugs.  What I mean is, for example, this entire feature isn't being tested as it has been black listed here: https://github.com/stackgl/headless-gl/blob/e558e9950e8c4acf1bf8a4e18c5c2af80cbfe636/test/extensions.js#L6

What I should have clarified is: Does that entire blacklisted suite need to pass before this lands?

**Comment by dhritzkiv** [comment]
btw, I think I found the source of the zeros:

https://github.com/stackgl/headless-gl/blob/f55a53347f966e5e14b01f9790ece627cf4464a9/test/multiple-render-targets.js#L39

**Comment by dhritzkiv** [comment]
> Does that entire blacklisted suite need to pass before this lands?

Good question. I'll have to think about that. My inclination is that the test should no longer be black listed if we're adding this functionality in. However, it may have been blacklisted for other reasons.

**Comment by robertleeplummerjr** [comment]
Then I'll get to work on unit tests asap.

**Comment by robertleeplummerjr** [comment]
Was failing 130 or so tests, now down to 18 with extension tests on, fyi.

File: webgl.js [hunk 5]
```diff
@@ -2794,6 +2831,9 @@ gl.getParameter = function getParameter (pname) {
           case ext.DRAW_BUFFER13_WEBGL:
           case ext.DRAW_BUFFER14_WEBGL:
           case ext.DRAW_BUFFER15_WEBGL:
+            if (ext._buffersState.length === 1 && ext._buffersState[0] === gl.BACK) {
+              return gl.BACK
+            }
```
Comment (dhritzkiv) [inline webgl.js]: Should the implicit `else` here break or return something specific?

**Comment by dhritzkiv** [comment]
Impressive. So very close!

File: webgl.js (hunk 5 above)
Comment (robertleeplummerjr) [reply webgl.js]: It should fall through.  I'll button that up before it is done.  Was in prototyping mode...

**Comment by robertleeplummerjr** [comment]
After digging for a day, I just tested failing tests 417 through 423, in Chrome, Firefox, and Safari, only to find that headless-gl works exactly like Chrome and Firefox, where the test is written for the functionality of Safari.

I believe these to be false positives.  I used a utility I wrote called gl-wiretap to output the logs, which I then put into an html file, which was where I noticed the deviation mentioned above here: https://gist.github.com/robertleeplummerjr/8461058f93c4652d9352b13cd1cd0268

The unit test seems to want all textures to get this new color.  Here is an image of the three browser together:

screen shot 2019-03-08 at 2 16 13 pm2


If this can be confirmed, we might want to change the unit test?  In any case, current passing unit tests are 16.  With this new math: `423-417=6;  16-6=10`,  possibly now at 10 failing unit tests within the extension.

**Comment by robertleeplummerjr** [comment]
Fixed a typo ^

**Comment by robertleeplummerjr** [comment]
Looks like tests 433, 434, and 435 are same.  Headless-gl in this PR working just like chrome and firefox, however the tests are written to how safari works.  To be clear what tests I'm talking about: `facet test/extensions.js` with this line commented out:

`// 'extensions_webgl-draw-buffers'`.

So it would seem: 10 - 3 = 7 failing unit tests

**Comment by robertleeplummerjr** [comment]
Html/javascript reproducing the issue: https://gist.github.com/robertleeplummerjr/f4ff59aa93704e6142489f2225aa1e70 (you have to step through the last few instruction sets, mostly this is a link for my future self).

**Comment by dhritzkiv** [comment]
Using the code in your [latest gist,](https://gist.github.com/robertleeplummerjr/f4ff59aa93704e6142489f2225aa1e70) I created a [code pen](https://codepen.io/dhritzkiv/pen/wOejQP?editors=1000#0)

Interestingly, Safari (Technology Preview) renders the final result the same as FF and Chrome.

**Comment by dhritzkiv** [comment]
Out of curiousity, do you have an idea of why headless-gl works exactly like Chrome and Firefox (and not Safari) w.r.t. draw-buffers extension? What exactly is the difference, do you believe, other than the final pixel output?

**Comment by dhritzkiv** [comment]
For your reference, if you haven't already used it, here are the conformance tests that are used in headless-gl, for the browser: https://www.khronos.org/registry/webgl/conformance-suites/1.0.3/webgl-conformance-tests.html

In running the webgl-draw-buffers extension, FF and Safari (TP) both fail, but Chrome passes. Curious.

**Comment by robertleeplummerjr** [comment]
>do you have an idea of why headless-gl works exactly like Chrome and Firefox

Yes, because I built it like that.  JK.  I've been tracing these bugs around looking for things that are broken, and have been comparing functionality to chrome and firefox, to ensure to match them as close as possible is all.  I thought about asking that question on freenode opengl, because I don't think I can find it documented anywhere.

>What exactly is the difference, do you believe, other than the final pixel output?

What the unit tests expects, I think, as indicated by the test name `debug("test that gl_FragColor broadcasts");` is that when using `gl_FragColor = something` the attached textures automatically receive this value as well.

To get it to work exactly like the unit tests: You'd have to manually trigger the update per render target rather than expecting it to propagate automatically for each texture.

**Comment by robertleeplummerjr** [comment]
Aside from that, I looked into handling the other failing tests, which have to do with the use of `#ifdef GL_EXT_draw_buffers` in code, with erroneous script, and the defining it if it isn't already, and I really think that is outside the scope of this issue.  I as well checked (painstakingly) the remaining failing extension tests, and they are duplicates in each case of the same issue stated above about differences in ff, chrome, and safari, and expecting safari like behavior.

With that being said, I believe the acceptable unit tests if we want this extension to mimic chrome and ff, are passing.  I did see a "preprocessor", which may be able to sort out `#ifdef GL_EXT_draw_buffers`, which I'm looking at now just to double check where we are.  Also, I've fixed lint (early switch return lint error) and as well added documention in readme.md.

**Comment by robertleeplummerjr** [comment]
>For your reference, if you haven't already used it, here are the conformance tests that are used in >headless-gl, for the browser: https://www.khronos.org/registry/webgl/conformance-suites/1.0.3/webgl-conformance-tests.html
>
>In running the webgl-draw-buffers extension, FF and Safari (TP) both fail, but Chrome passes. Curious.

FF passes fine for me!  There has to be something I'm missing then.

**Comment by robertleeplummerjr** [comment]
Ok, I don't think it (broadcasting the color from gl_FragColor) is supported in the version of google angle we are on.

The missing feature implementation from google angle: https://github.com/google/angle/blob/7c37ca1836402b654dbbf5f00007e09bdf2a7051/src/compiler/translator/tree_ops/EmulateGLFragColorBroadcast.cpp

Specifically: https://github.com/google/angle/blob/7c37ca1836402b654dbbf5f00007e09bdf2a7051/src/compiler/translator/tree_ops/EmulateGLFragColorBroadcast.cpp#L114

We could shim in something like this, but ultimately I'd like to keep our stuff as vanilla as possible.

Neither the feature or the unit test are in our version, fyi.

**Comment by dhritzkiv** [comment]
Yes, let's keep it as-is and not shim it. Hopefully in the future if/when we upgrade the angle implementation, we can expect behaviour more consistent with the spec and have those tests pass.

---

Other than the  `#ifdef GL_EXT_draw_buffers` issue, and the cross-platform implementation differences, is there anything else keeping the draw-buffer extension suite from passing? We can skip the draw-buffer extension suite for this PR/release, ensure all other tests pass, and proceed with final reviews before merging.

---

Also, I believe the multiple render targets test you added needs some fixing up to have the correct expected vs. actual zeros quantities.

**Comment by robertleeplummerjr** [comment]
>Other than the #ifdef GL_EXT_draw_buffers issue, and the cross-platform implementation differences, is there anything else keeping the draw-buffer extension suite from passing? We can skip the draw-buffer extension suite for this PR/release, ensure all other tests pass, and proceed with final reviews before merging.

I don't think so, (also, already built an upgrade script for google angle for later on down the road, he he) looking into other tests and the new test I added.  Ty @dhritzkiv !

**Comment by robertleeplummerjr** [comment]
I run into some funny business with the following tests:
```
["more_functions_copyTexImage2DBadArgs", require("./more_functions_copyTexImage2DBadArgs")],

["more_functions_copyTexSubImage2D", require("./more_functions_copyTexSubImage2D")],  // This one is blacklisted, should the others?

["more_functions_copyTexSubImage2DBadArgs", require("./more_functions_copyTexSubImage2DBadArgs")],
```

They cause a `Abort trap: 6`, but if I comment them out, or blacklist them, I now get all passing unit tests.

File: test/multiple-render-targets.js [hunk 6]
```diff
@@ -20,13 +20,13 @@ tape('multiple-render-targets', function (t) {
       notZero++
     }
   }
-  t.equals(notZero, 144000, `Only ${notZero} are not 0, expected 144000`)
+  t.equals(notZero, 15, `Only ${notZero} are not 0, expected 15`)
```
Comment (dhritzkiv) [inline test/multiple-render-targets.js]: Tests are now passing!

btw, what's the rationale/math behind `15`?

**Comment by dhritzkiv** [comment]
Hmm, I don't have an issue with those suites on my machine…

**Comment by robertleeplummerjr** [comment]
It is probably either my node version (10) or operating system (osx 10.13.3 (17D47)).  

Regards "15".  I forgot to account for the missing pixel you pointed out, the one that is "0".  Fixing it, will have it updated momentarily.

File: test/multiple-render-targets.js (hunk 6 above)
Comment (robertleeplummerjr) [reply test/multiple-render-targets.js]: fixed to be 16, sorry about that.

**Comment by dhritzkiv** [comment]
Hmm. I'm running the exact same environment. I imagine our GFX cards are different (but that also shouldn't matter). Mine is the NVIDIA GeForce GTX 675MX 1024 MB (from 2011, 2012)

**Comment by robertleeplummerjr** [comment]
I'm running the Intel stock, don't know the details this moment.

**Comment by robertleeplummerjr** [comment]
Regardless, it is a separate issue to those raised by this PR.  What needs to happen next with this PR to be merge worthy?

**Comment by robertleeplummerjr** [comment]
My GFX is an: Intel Iris Pro 1536 MB

**Comment by dhritzkiv** [comment]
I'll do at least one more review of this PR's code, tests, and README over the next few days, and I'll invite the other maintainer of the repo to do so as well. Overall, it's looking to be in a good place.

After that, we'll merge into master, and then cut a new release for npm (along with generating prebuilt files for easy install).

**Comment by dhritzkiv** [comment]
@mourner Hey! Could you look over this PR quick? I need to do another pass over soon, myself, but overall it looks good.

Summary of the state of this:
- New extension!
- Existing tests pass (locally†)
- New multiple-render targets test testing the extension also passes
- Un-blacklisting the webgl-draw-buffers extension test suite leads to a small handful of assertion tests tests failing, but it's not for a lack of trying: there seem to be implementation differences between the three major browsers themselves, as well as the current version of angle that we're using in headless-gl, as well as graphics card of the user machine. In the end, these differences are what I'd consider edge-case. Ultimately, we can simply keep that suite blacklisted and we'd still get an all-green.††

† Travis-CI is now completely non-functional for this repo, as the connection with Travis-CI is now done through Github integrations/apps. Migrating to the new integration method requires ownership access to the repo. I will contact stackgl personally and make a plea for them to grant access to you and/or me.

†† my rationale for continuing to ignore the webgl-draw-buffers extension suite: 1) this is a non-breaking change; 2) the tests for that extension never worked and are outdated/diverged from the Khronos conformance tests suite (which themselves don't pass in every browser).

---

Robert here has been working super hard on this PR in order to support his awesome [gpujs/gpu.js](https://github.com/gpujs/gpu.js) project and by extension [BrainJS/brain.js](https://github.com/BrainJS/brain.js). Both highly successful and impressive works.

He's been working on this since well before this PR's creation. He and I have been talking behind the scenes and he's expressed an interest in becoming a maintainer on this repo. Given his excellent work on this PR, his authorship of the two aforementioned products, and his overall enthusiasm for this project, I would recommend him as a maintainer. Robert also has started on other potential headless-gl improvements, such as updating angle which includes the support of webgl-2.

**Comment by dhritzkiv** [comment]
@robertleeplummerjr FYI: I added a small stylistic commit to the PR in order to trigger a CI build

**Comment by dhritzkiv** [comment]
I'll wait until midday tomorrow for the other maintainer to chime in. After that, I'll merge, version bump, and release on npm.

**Comment by mourner** [comment]
@robertleeplummerjr Just invited you as a collaborator. Thanks again for taking on this — I'm excited!

**Comment by robertleeplummerjr** [comment]
Ty guys.  Invite accepted!

**Pull request merged**